src/rust/runtime/network/ephemeral.rs

```diff
@@ -0,0 +1,246 @@
+
+const FIRST_PRIVATE_PORT: u16 = 49152;
+const LAST_PRIVATE_PORT: u16 = 65535;
+/// Seed number of ephemeral port allocator.
+const EPHEMERAL_PORT_SEED: u64 = 12345;
+
+//======================================================================================================================
+// Structures
+//======================================================================================================================
+
+pub struct EphemeralPorts {
+    ports: Vec<u16>,
+}
+
+//======================================================================================================================
+// Associated Functions
+//======================================================================================================================
+
+impl EphemeralPorts {
+    /// Creates a new ephemeral port allocator.
+    pub fn new(rng: &mut SmallRng) -> Self {
+        let mut ports: Vec<u16> = Vec::<u16>::new();
+        for port in FIRST_PRIVATE_PORT..LAST_PRIVATE_PORT {
+            ports.push(port);
+        }
+        ports.shuffle(rng);
+        Self { ports }
+    }
+
+    /// Asserts wether a port is in the ephemeral port range.
+    pub fn is_private(port: u16) -> bool {
+        port >= FIRST_PRIVATE_PORT
+    }
+
+    /// Allocates any ephemeral port from the pool.
+    pub fn alloc(&mut self) -> Result<u16, Fail> {
+        self.ports.pop().ok_or(Fail::new(
+            libc::EADDRINUSE,
+            "all port numbers in the ephemeral port range are currently in use",
+        ))
+    }
+
+    /// Allocates the specified port from the pool.
+    pub fn reserve(&mut self, port: u16) -> Result<(), Fail> {
+        // Check if port is not in the pool.
+        if !self.ports.contains(&port) {
+            return Err(Fail::new(libc::ENOENT, "port number not found"));
+        }
+
+        // Remove port from the pool.
+        self.ports.retain(|&p| p != port);
+
+        Ok(())
+    }
+
+    /// Releases a ephemeral port.
+    pub fn free(&mut self, port: u16) -> Result<(), Fail> {
+        // Check if port is in the valid range.
+        if !Self::is_private(port) {
+            let cause: String = format!("port {} is not in the ephemeral port range", port);
+            error!("free(): {}", &cause);
+            return Err(Fail::new(libc::EINVAL, &cause));
+        }
+
+        // Check if port is already in the pool.
+        if self.ports.contains(&port) {
+            let cause: String = format!("port {} is already in the pool", port);
+            error!("free(): {}", &cause);
+            return Err(Fail::new(libc::EFAULT, &cause));
+        }
+
+        self.ports.push(port);
+
+        Ok(())
+    }
+}
+
+//======================================================================================================================
+// Trait Implementations
+//======================================================================================================================
+
+impl Default for EphemeralPorts {
```

Comment:
There are multiple identical implementations for EphemeralPorts. Is it possible to reuse?

Reply:
Ah, I've deleted it now.